src/serenity/serenity.hpp

```diff
@@ -12,121 +12,205 @@
 namespace mesos {
 namespace serenity {
 
+class BaseFilter {
+  template <typename T>
+  friend class Consumer;
+  template <typename T>
+  friend class Producer;
+ protected:
+  /**
+   * Function invoked when filter consumes all products in iteration.
+   * When filter consumes more than one product, this function should
+   * be overriden.
+   *
+   * TODO(skonefal): After consume deprecation, this should be abstract.
```

Comment:
We should remove _consume_ from our components and replace it with virtual abstract _allProductsReady_ to keep consistence. What do you think?